tiledb/dense_array.py

```diff
@@ -0,0 +1,785 @@
+        self, subarray, attr_names: list, cond, layout, include_coords
+    ):
+        from .main import PyQuery
+
+        q = PyQuery(self._ctx_(), self, tuple(attr_names), tuple(), layout, False)
+        self.pyquery = q
+
+        if cond is not None and cond != "":
+            if isinstance(cond, str):
+                from .query_condition import QueryCondition
+
+                q.set_cond(QueryCondition(cond))
+            else:
+                raise TypeError("`cond` expects type str.")
+
+        q.set_subarray(subarray)
+        q.submit()
+        results = OrderedDict()
+        results = q.results()
+
+        out = OrderedDict()
+
+        output_shape = subarray.shape()
+
+        nattr = len(attr_names)
+        for i in range(nattr):
+            name = attr_names[i]
+            if not self.schema.domain.has_dim(name) and self.schema.attr(name).isvar:
+                # for var arrays we create an object array
+                dtype = object
+                out[name] = q.unpack_buffer(
+                    name, results[name][0], results[name][1]
+                ).reshape(output_shape)
+            else:
+                dtype = q.buffer_dtype(name)
+
+                # <TODO> sanity check the TileDB buffer size against schema?
+                # <TODO> add assert to verify np.require doesn't copy?
+                arr = results[name][0]
+                arr.dtype = dtype
+                if len(arr) == 0:
+                    # special case: the C API returns 0 len for blank arrays
+                    arr = np.zeros(output_shape, dtype=dtype)
+                elif len(arr) != np.prod(output_shape):
+                    raise Exception(
+                        "Mismatched output array shape! (arr.shape: {}, output.shape: {}".format(
+                            arr.shape, output_shape
+                        )
+                    )
+
+                if layout == lt.LayoutType.ROW_MAJOR:
+                    arr.shape = output_shape
+                    arr = np.require(arr, requirements="C")
+                elif layout == lt.LayoutType.COL_MAJOR:
+                    arr.shape = output_shape
+                    arr = np.require(arr, requirements="F")
+                else:
+                    arr.shape = np.prod(output_shape)
+
+                out[name] = arr
+
+            if self.schema.has_attr(name) and self.attr(name).isnullable:
+                out[name] = np.ma.array(out[name], mask=~results[name][2].astype(bool))
+
+        return out
+
+    def __setitem__(self, selection, val):
+        """Set / update dense data cells
+
+        :param tuple selection: An int index, slice or tuple of integer/slice objects,
+            specifiying the selected subarray region for each dimension of the DenseArray.
+        :param value: a dictionary of array attribute values, values must able to be converted to n-d numpy arrays.\
+            if the number of attributes is one, then a n-d numpy array is accepted.
```

Comment:
nit: Match doc to arg name

```suggestion
    def __setitem__(self, selection, val):
        """Set / update dense data cells

        :param tuple selection: An int index, slice or tuple of integer/slice objects,
            specifiying the selected subarray region for each dimension of the DenseArray.
        :param val: a dictionary of array attribute values, values must able to be converted to n-d numpy arrays.\
            if the number of attributes is one, then a n-d numpy array is accepted.
```